server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/PropStoreWatcher.java

```diff
@@ -0,0 +1,251 @@
+
+  private static final ExecutorService executorService =
+      ThreadPools.getServerThreadPools().createFixedThreadPool(2, "zoo_change_update", false);
+
+  private final ReentrantReadWriteLock listenerLock = new ReentrantReadWriteLock();
+  private final ReentrantReadWriteLock.ReadLock listenerReadLock = listenerLock.readLock();
+  private final ReentrantReadWriteLock.WriteLock listenerWriteLock = listenerLock.writeLock();
+
+  // access should be guarded by acquiring the listener read or write lock
+  private final Map<PropCacheKey,Set<PropChangeListener>> listeners = new HashMap<>();
+
+  private final ReadyMonitor zkReadyMonitor;
+
+  public PropStoreWatcher(final ReadyMonitor zkReadyMonitor) {
+    this.zkReadyMonitor = zkReadyMonitor;
+  }
+
+  public void registerListener(final PropCacheKey propCacheKey, final PropChangeListener listener) {
+    listenerWriteLock.lock();
+    try {
+      Set<PropChangeListener> set = listeners.computeIfAbsent(propCacheKey, s -> new HashSet<>());
+      set.add(listener);
+    } finally {
+      listenerWriteLock.unlock();
+    }
+  }
+
+  /**
+   * Process a ZooKeeper event. This method does not reset the watcher. Subscribers are notified of
+   * the change - if they call get to update and respond to the change the watcher will be (re)set
+   * then. This helps clean up watchers by not automatically re-adding the watcher on the event but
+   * only if being used.
+   *
+   * @param event
+   *          ZooKeeper event.
+   */
+  @Override
+  public void process(final WatchedEvent event) {
+
+    String path;
+    PropCacheKey propCacheKey;
+    switch (event.getType()) {
+      case NodeDataChanged:
+        path = event.getPath();
+        log.trace("handle change event for path: {}", path);
+        propCacheKey = PropCacheKey.fromPath(path);
+        if (propCacheKey != null) {
+          signalZkChangeEvent(propCacheKey);
+        }
+        break;
+      case NodeDeleted:
+        path = event.getPath();
+        log.trace("handle delete event for path: {}", path);
+        propCacheKey = PropCacheKey.fromPath(path);
+        if (propCacheKey != null) {
+          // notify listeners
+          Set<PropChangeListener> snapshot = getListenerSnapshot(propCacheKey);
+          if (snapshot != null) {
+            executorService
+                .execute(new PropStoreEventTask.PropStoreDeleteEventTask(propCacheKey, snapshot));
+          }
+          listenerCleanup(propCacheKey);
+        }
+        break;
+      case None:
+        Event.KeeperState state = event.getState();
+        switch (state) {
+          // pause - could reconnect
+          case ConnectedReadOnly:
+          case Disconnected:
+            log.debug("ZooKeeper disconnected event received");
+            zkReadyMonitor.clearReady();
+            executorService.execute(
+                new PropStoreEventTask.PropStoreConnectionEventTask(getAllListenersSnapshot()));
```

Comment:
I did not think that the code would be sensitive to order. All events basically signal the cache to clear the entry.  The cache is not trying to respond to each change, it will fetch a changed value on demand when / if the next read occurs.  This means that the event order for changes should not matter.

In the case of a disconnect - the prop store watcher first sets ready to false so any reads because changes will block until ready. And that occurs independent of the notification of the changes. Even if there are change notifications queued and in-flight, any reads will be blocked. 

In the case of multiple events and reads while other events are in-flight  Each event may just re-clear a cleared entry - so basically a no-op.

(below, current - final is the stabilized "final" value, until changed again by future updates")

In the cases where a clear / read occurs and there are additional events in the pipeline each change could trigger a clear and then a potential read. The read may not get the values that were set with that specific change event, but it will get a value from that event or a value set later. If the value was as an intermediate change, subsequent events should re-clear and trigger reads until the current, final value is read.  Once it has the current, final value, events in the pipeline may be clearing / re-reading the same "current - final"  value, but the final state of the event chain should always end up with the correct, current state.